File: LeetCode/0001. Two Sum/solution.cpp
#include <unordered_map>
/**
 * 19 / 19 test cases passed.
 * Status: Accepted
 * Runtime: 8ms
 */
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> dic;
        for (int i = 0; i < nums.size(); i++) {
            int val = target - nums[i];
            if (dic.find(val) != dic.end()) {
                vector<int> res;
                res.push_back(dic[val]);
                res.push_back(i);
                return res;
            } else dic[nums[i]] = i;
        }
        return vector<int>();
    }
};

File: LeetCode/0001. Two Sum/solution.rs
use std::collections::HashMap;

/**
 * 57 / 57 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 2.5 MB 
 */
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut cache = HashMap::new();
        for (i, num) in nums.iter().enumerate() {
            if let Some(n) = cache.get(&(target - num)) {
                return vec![*n, i as i32];
            }
            cache.insert(num, i as i32);
        }
        vec![]
    }
}
